### PR TITLE
Support multiple vectors for a single target

### PR DESCRIPTION
**Pull request by dirkkul**

### CODE REVIEW

File: weaviate/collections/aggregations/hybrid.py [hunk 1]
```diff
@@ -25,6 +25,7 @@ async def hybrid(
         filters: Optional[_Filters] = None,
         group_by: Optional[Union[str, GroupByAggregate]] = None,
         target_vector: Optional[str] = None,
+        max_vector_distance: Optional[float] = None,
```
Comment (tsmith023) [inline weaviate/collections/aggregations/hybrid.py:28]: I noticed that this is `Optional[NUMBER]` in the `.queries` methods. Should we align these or is this a point-of-difference due to the differing under-the-hood protocols?

File: integration/test_named_vectors.py [hunk 2]
```diff
@@ -641,3 +634,120 @@ def test_multi_query_error_no_target_vector(collection_factory: CollectionFactor
...
+    ],
+)
+def test_same_target_vector_multiple_input(
+    collection_factory: CollectionFactory,
+    target_vector: Union[List[str], _MultiTargetVectorJoin],
+    distances: List[float],
+) -> None:
+    dummy = collection_factory("dummy")
+    # Todo: Update this to 1.26.2
+    if dummy._connection._weaviate_version.is_lower_than(1, 26, 1):
```
Comment (tsmith023) [inline integration/test_named_vectors.py:660]: Does this `todo` need addressing?

File: integration/test_named_vectors.py [hunk 3]
```diff
@@ -641,3 +634,120 @@ def test_multi_query_error_no_target_vector(collection_factory: CollectionFactor
...
+        ),
+    ],
+)
+def test_same_target_vector_multiple_input_combinations(
+    collection_factory: CollectionFactory,
+    near_vector: Dict[str, Union[Sequence[float], Sequence[Sequence[float]]]],
+    target_vector: List[str],
+) -> None:
+    dummy = collection_factory("dummy")
+    # Todo: Update this to 1.26.2
```
Comment (tsmith023) [inline integration/test_named_vectors.py:701]: Does this `todo` need addressing?

File: integration/test_named_vectors.py [hunk 4]
```diff
@@ -641,3 +634,120 @@ def test_multi_query_error_no_target_vector(collection_factory: CollectionFactor
...
+
+    objs = collection.query.near_vector(
+        near_vector, target_vector=target_vector, return_metadata=wvc.query.MetadataQuery.full()
+    ).objects
+    assert sorted([obj.uuid for obj in objs]) == sorted([uuid2, uuid1])
+
+
+def test_deprecated_syntax(collection_factory: CollectionFactory):
+    dummy = collection_factory("dummy")
+    # Todo: Update this to 1.26.2
```
Comment (tsmith023) [inline integration/test_named_vectors.py:724]: Does this `todo` need addressing?

File: weaviate/collections/grpc/query.py [hunk 5]
```diff
@@ -235,23 +235,53 @@ def hybrid(
...
-                vector_per_target=vector_per_target,
+                vector_per_target=vector_per_target_tmp,
+                vector_for_targets=vector_for_targets_tmp,
             )
         else:
-            vector_per_target, vector_bytes_tmp = self.__vector_per_target(
-                vector, targets, "vector"
-            )
-            if vector_per_target is not None:
+            # Todo: increase to .2
```
Comment (tsmith023) [inline weaviate/collections/grpc/query.py:262]: Is this TODO still relevant?

File: weaviate/collections/grpc/query.py [hunk 6]
```diff
@@ -382,20 +412,30 @@ def near_vector(
...
             near_vector_grpc: Optional[bytes] = struct.pack(
                 "{}f".format(len(near_vector)), *near_vector
             )
             vector_per_target_tmp = None
+            vector_for_targets = None
         else:
-            vector_per_target_tmp, near_vector_grpc = self.__vector_per_target(
-                near_vector, targets, "near_vector"
-            )
+            # Todo: increase to .2
```
Comment (tsmith023) [inline weaviate/collections/grpc/query.py:423]: Is this TODO still relevant?

File: integration/test_collection_hybrid.py [hunk 7]
```diff
@@ -384,14 +383,10 @@ def test_hybrid_near_text_search_named_vectors(collection_factory: CollectionFac
...
+        # {"first": [1, 0], "second": [1, 0, 0]},
         {"first": [1, 0], "second": np.array([1, 0, 0])},
-        {"first": [1, 0], "second": pl.Series([1, 0, 0])},
-        {"first": [1, 0], "second": pd.Series([1, 0, 0])},
-        [[1, 0], [1, 0, 0]],
-        [[1, 0], np.array([1, 0, 0])],
-        [[1, 0], pl.Series([1, 0, 0])],
-        [[1, 0], pd.Series([1, 0, 0])],
+        # {"first": [1, 0], "second": pl.Series([1, 0, 0])},
+        # {"first": [1, 0], "second": pd.Series([1, 0, 0])},
```
[the remark this answers is not in the document]
Comment (dirkkul) [reply integration/test_collection_hybrid.py]: commented out for testing and then forgot, thx!

File: integration/test_named_vectors.py [hunk 8]
```diff
@@ -641,3 +634,126 @@ def test_multi_query_error_no_target_vector(collection_factory: CollectionFactor
...
+        ),
+    ],
+)
+def test_same_target_vector_multiple_input_combinations(
+    collection_factory: CollectionFactory,
+    near_vector: Dict[str, Union[Sequence[float], Sequence[Sequence[float]]]],
+    target_vector: List[str],
+) -> None:
+    dummy = collection_factory("dummy")
+    # Todo: Update this to 1.26.2
```
[the remark this answers is not in the document]
Comment (dirkkul) [reply integration/test_named_vectors.py]: no, updated the numbers and forgot the Todo

File: weaviate/collections/grpc/query.py [hunk 9]
```diff
@@ -921,20 +975,67 @@ def __vector_per_target(
...
-                    ):
-                        raise invalid_nv_exception
-                    vector_per_target[targets.target_vectors[i]] = struct.pack(
-                        "{}f".format(len(nv)), *nv
-                    )
-                return vector_per_target, None
+                raise WeaviateInvalidInputError(
+                    """Providing lists of lists has been deprecated. Please provide a dictionary with target names as
+                    keys and lists of numbers as values."""
+                )
```
[the remark this answers is not in the document]
Comment (dirkkul) [reply weaviate/collections/grpc/query.py:978]: yes, this is breaking.

Discussed this a while ago and:
- it wasn't present for that long, so tehre should be very few users
- not aprt of the docs, even less users
- with multi vector that syntax gets too confusing and if we kept the old behaviour you could do single vector multi target search but not multi vector.

not super happy about this, but seemed like the best solution

[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/grpc/query.py:978]: > not part of the docs, even less users

I think this is a good enough reason to break then!

**Review by tsmith023 — APPROVED** [review]

**Pull request merged**